api/credentials_controller.go

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2018 The Service Manager Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package api
+
+import (
+	"context"
+	"fmt"
+	"github.com/Peripli/service-manager/api/osb"
+	"github.com/Peripli/service-manager/pkg/log"
+	"github.com/Peripli/service-manager/pkg/query"
+	"github.com/Peripli/service-manager/pkg/util"
+	"github.com/gofrs/uuid"
+	"github.com/tidwall/gjson"
+	"net/http"
+	"time"
+
+	"github.com/Peripli/service-manager/pkg/types"
+	"github.com/Peripli/service-manager/pkg/web"
+)
+
+// CredentialsController implements api.Controller by providing logic for broker platform credential storage/update
+type CredentialsController struct {
+	*BaseController
+}
+
+func NewCredentialsController(ctx context.Context, options *Options) *CredentialsController {
+	return &CredentialsController{
+		BaseController: NewController(ctx, options, web.BrokerPlatformCredentialsURL, types.BrokerPlatformCredentialType, func() types.Object {
+			return &types.BrokerPlatformCredential{}
+		}),
+	}
+}
+
+func (c *CredentialsController) Routes() []web.Route {
+	return []web.Route{
+		{
+			Endpoint: web.Endpoint{
+				Method: http.MethodPut,
+				Path:   c.resourceBaseURL,
+			},
+			Handler: c.setCredentials,
+		},
+	}
+}
+func (c *BaseController) setCredentials(r *web.Request) (*web.Response, error) {
+	ctx := r.Context()
+	log.C(ctx).Debugf("Creating new broker platform credentials")
+
+	platform, err := osb.ExtractPlatformFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	body := &types.BrokerPlatformCredential{}
+	if err := util.BytesToObject(r.Body, body); err != nil {
+		return nil, err
+	}
+	body.PlatformID = platform.ID
+
+	if body.NotificationID != "" {
+		criteria := []query.Criterion{
+			query.ByField(query.EqualsOperator, "id", body.NotificationID),
+			query.ByField(query.EqualsOrNilOperator, "platform_id", body.PlatformID),
+			query.ByField(query.EqualsOperator, "resource", string(types.ServiceBrokerType)),
+		}
+
+		obj, err := c.repository.Get(ctx, types.NotificationType, criteria...)
+		if err != nil {
+			if err == util.ErrNotFoundInStorage {
+				return nil, &util.HTTPError{
+					ErrorType:   "CredentialsError",
+					Description: fmt.Sprintf("Invalid notification ID %s - request rejected", body.NotificationID),
+					StatusCode:  http.StatusConflict,
```

Comment:
Bad request/Unauthorized are pretty common request errors - we wanted to make sure that we return something more unique, so that the proxy client could interpret it and know how to handle it and know that it's a "happy path" scenario instead of some real issue. Also the idea for Conflict was that the credentials are already in SM DB.

Reply:
my understanding was that full resync wont do credentials rotation as the flow is not initiated by sm?